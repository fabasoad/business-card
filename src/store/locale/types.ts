export const LOAD_LOCALE = 'LOAD_LOCALE'
export const SET_LOCALE = 'SET_LOCALE'

export class Locale {
  code: string
  title: string

  constructor(code: string, title: string = null) {
    this.code = code
    this.title = ` ${title || code}`
  }
}

interface LoadLocaleAction {
  type: typeof LOAD_LOCALE
  payload: Locale
}

interface SetLocaleAction {
  type: typeof SET_LOCALE
  payload: Locale
}

export type LocaleActionTypes = LoadLocaleAction | SetLocaleAction

export type AppActions = LocaleActionTypes
