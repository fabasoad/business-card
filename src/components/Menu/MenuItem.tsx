import * as React from 'react'
import { Nav } from 'react-bootstrap'
import { useTranslation } from 'react-i18next'

interface MenuItemProps {
  name: string
  setActiveNavLink: React.Dispatch<React.SetStateAction<string>>
}

export default function MenuItem(props: MenuItemProps) {
  const { t } = useTranslation()
  return (
    <Nav.Link
      onClick={() => props.setActiveNavLink(`#${props.name}`)}
      eventKey={`#${props.name}`}
      href={`#${props.name}`}
    >
      {t(`business-card-menu-${props.name}`)}
    </Nav.Link>
  )
}
