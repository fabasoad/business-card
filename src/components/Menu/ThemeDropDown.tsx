import * as React from 'react'
import { Dropdown } from 'react-bootstrap'
import { useTranslation } from 'react-i18next'
import ThemeContext, { SUPPORTED_THEMES } from '../Contexts/ThemeContext'

export default function ThemeDropDown() {
  const { t } = useTranslation()
  const { theme, setTheme } = React.useContext(ThemeContext)

  return (
    <Dropdown>
      <Dropdown.Toggle bsPrefix="nav-link dropdown-toggle" variant={null}>
        <i className="fa fa-solid fa-circle-half-stroke"></i>
        {t(`themes.${theme}`)}
      </Dropdown.Toggle>
      <Dropdown.Menu>
        {SUPPORTED_THEMES.filter(v => v !== theme).map(v => {
          return (
            <Dropdown.Item
              onClick={() => setTheme(v)}
              bsPrefix="nav-link"
              eventKey={v}
              key={`dropdown-item-theme-${v}`}
            >
              {t(`themes.${v}`)}
            </Dropdown.Item>
          )
        })}
      </Dropdown.Menu>
    </Dropdown>
  )
}
