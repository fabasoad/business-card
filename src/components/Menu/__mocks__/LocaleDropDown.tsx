import * as React from 'react'

export default function LocaleDropDown() {
  return (
    <div data-testid="LocaleDropDown" />
  )
}
