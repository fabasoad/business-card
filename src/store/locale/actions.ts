import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import { Action } from 'redux'
import { AppState } from '../configureStore'
import { ThunkAction } from 'redux-thunk'
import { Dispatch } from 'react'
import { AppActions, Locale } from './types'
import SupportedLocales from '../../scripts/SupportedLocales'
import * as gb from './resources/gb.json'
import * as jp from './resources/jp.json'
import * as ru from './resources/ru.json'

type ThunkLocaleAction = ThunkAction<void, AppState, unknown, Action<string>>

const onLanguageChange = (t) => {
  document.title = t('business-card-title')
}

export const loadLocale = (payload: Locale): AppActions => ({
  type: 'LOAD_LOCALE',
  payload
})

export const startLoadLocale = (): ThunkLocaleAction =>
  async (dispatch: Dispatch<AppActions>) => {
    const params = new URLSearchParams(document.location.search)
    const code: string = params.get('locale') || localStorage.getItem('i18nextLng')
    const currentLocale = SupportedLocales.getOrDefault(code)
    if (document.location.search) {
      window.history.replaceState(
        {},
        document.title,
        document.location.origin + document.location.pathname + document.location.hash
      )
    }
    await i18n
      .use(LanguageDetector)
      .use(initReactI18next)
      .init({
        debug: process.env.NODE_ENV === 'development',
        defaultNS: 'common',
        fallbackLng: SupportedLocales.default.code,
        lng: currentLocale.code,
        ns: ['common'],
        resources: {
          gb: { common: gb },
          jp: { common: jp },
          ru: { common: ru }
        }
      }).then((t) => {
        onLanguageChange(t)
        dispatch(
          loadLocale({
            ...currentLocale
          })
        )
      })
  }

export const setLocale = (payload: Locale): AppActions => ({
  type: 'SET_LOCALE',
  payload
})

export const startSetLocale = (locale: Locale): ThunkLocaleAction =>
  (dispatch: Dispatch<AppActions>) => {
    i18n.changeLanguage(locale.code, (err, t) => {
      onLanguageChange(t)
      dispatch(
        setLocale({
          ...locale
        })
      )
    })
  }
