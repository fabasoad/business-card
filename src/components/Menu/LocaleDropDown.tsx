import * as React from 'react'
import { Dropdown } from 'react-bootstrap'
import FlagIconFactory from 'react-flag-icon-css'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { ThunkDispatch } from 'redux-thunk'

import SupportedLocales from '../../scripts/SupportedLocales'
import { Locale, AppActions } from '../../store/locale/types'
import { AppState } from '../../store/configureStore'
import { startSetLocale } from '../../store/locale/actions'

interface LocaleDropDownProps {
  getLocalesExceptOf: (code: string) => Locale[]
}

type Props = LocaleDropDownProps & LinkDispatchProps & LinkStateProps

export function LocaleDropDown(props: Props) {
  const handleClick = (locale: Locale): void => {
    props.startSetLocale(locale)
  }
  const FlagIcon = FlagIconFactory(React, { 'useCssModules': false })
  return (
    <Dropdown>
      <Dropdown.Toggle bsPrefix="nav-link dropdown-toggle" variant={null} id="btnLocale">
        <FlagIcon code={props.locale.code} />
        {props.locale.title}
      </Dropdown.Toggle>
      <Dropdown.Menu>
        {props.getLocalesExceptOf(props.locale.code).map((l) => {
          return (
            <Dropdown.Item onSelect={() => handleClick(l)} bsPrefix="nav-link" eventKey={l.code} key={l.code}>
              <FlagIcon code={l.code} />
              {l.title}
            </Dropdown.Item>
          )
        })}
      </Dropdown.Menu>
    </Dropdown>
  )
}

LocaleDropDown.defaultProps = {
  getLocalesExceptOf: SupportedLocales.getExceptOf
}

interface LinkStateProps {
  locale: Locale
}

interface LinkDispatchProps {
  startSetLocale: (locale: Locale) => void
}

const mapStateToProps = (
  state: AppState,
  ownProps: LocaleDropDownProps
): LinkStateProps => ({
  locale: state.locale
})

const mapDispatchToProps = (
  dispatch: ThunkDispatch<any, any, AppActions>,
  ownProps: LocaleDropDownProps
): LinkDispatchProps => ({
  startSetLocale: bindActionCreators(startSetLocale, dispatch)
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LocaleDropDown)
