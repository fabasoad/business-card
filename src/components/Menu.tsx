import * as React from 'react';
import { useTranslation } from 'react-i18next';

import LocaleDropDown from './LocaleDropDown';

export default function Menu(props) {
  const { t, i18n } = useTranslation();

  return (
    <div className="affix-top" id="nav">
      <nav className="navbar navbar-expand-md navbar-custom">
        <a className="navbar-brand" href="#header">{t('business-card-author-name')}</a>
        <button className="navbar-toggler" id="dfnlas1u0" type="button" data-toggle="collapse" data-target=".navbar-container" aria-expanded="false"><i className="fa fa-bars"></i></button>
        <div
          className="navbar-container collapse navbar-collapse justify-content-between" aria-labelledby="dfnlas1u0">
          <ul className="navbar-nav">
            <li className="nav-item" data-toggle="collapse" data-target=".navbar-container.show">
              <a className="nav-link" href="#about">{t('business-card-menu-about-me')}</a>
            </li>
            <li className="nav-item" data-toggle="collapse" data-target=".navbar-container.show">
              <a className="nav-link" href="#experience">{t('business-card-menu-experience')}</a>
            </li>
            <li className="nav-item" data-toggle="collapse" data-target=".navbar-container.show">
              <a className="nav-link" href="#education">{t('business-card-menu-education')}</a>
            </li>
            <li className="nav-item" data-toggle="collapse" data-target=".navbar-container.show">
              <a className="nav-link" href="#portfolio">{t('business-card-menu-portfolio')}</a>
            </li>
            <li className="nav-item" data-toggle="collapse" data-target=".navbar-container.show">
              <a className="nav-link" href="#resume">{t('business-card-menu-resume')}</a>
            </li>
            <li className="nav-item" data-toggle="collapse" data-target=".navbar-container.show">
              <a className="nav-link" href="#contact">{t('business-card-menu-contact')}</a>
            </li>
          </ul>
          <ul className="navbar-nav">
            <li className="nav-item dropdown">
              <LocaleDropDown setLocale={props.setLocale} getLocale={props.getLocale} />
            </li>
          </ul>
        </div>
      </nav>
    </div>
  );
}