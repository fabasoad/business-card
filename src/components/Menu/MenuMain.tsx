import * as React from 'react'
import { Navbar, Nav } from 'react-bootstrap'
import { useTranslation } from 'react-i18next'

import LocaleDropDown from './LocaleDropDown'
import MenuItem from './MenuItem'

export default function MenuMain() {
  const { t } = useTranslation()

  const [expanded, setExpanded] = React.useState<boolean>(false)
  const [activeNavLink, setActiveNavLink] = React.useState<string>(document.location.hash)

  document.querySelectorAll('*').forEach((el) => {
    el.addEventListener('click', (e) => {
      const findParent = (el1: HTMLElement, el2: HTMLElement): HTMLElement => {
        if (el1 === el2) {
          return el1
        }
        if (el1.parentElement == null) {
          return null
        }
        return findParent(el1.parentElement, el2)
      }
      const element = e.target as HTMLElement
      const isNotNav = element.getAttribute('id') != 'nav'
      const isNotParentOfNav = findParent(document.getElementById('nav'), element) == null
      const isNavIsNotParent = findParent(element, document.getElementById('nav')) == null

      if (isNotNav && isNotParentOfNav && isNavIsNotParent) {
        setExpanded(false)
      }
    })
  })

  return (
    <div className="affix-top" id="nav">
      <Navbar expanded={expanded} onToggle={setExpanded} collapseOnSelect={true} expand="md" className="navbar-custom">
        <Navbar.Brand href="#header">{t('business-card-author-name')}</Navbar.Brand>
        <Navbar.Toggle className="navbar-dark" aria-controls="basic-navbar-nav" />
        <Navbar.Collapse id="basic-navbar-nav">
          <Nav activeKey={activeNavLink} className="mr-auto">
            <MenuItem setActiveNavLink={setActiveNavLink} name={'about'} />
            <MenuItem setActiveNavLink={setActiveNavLink} name={'skills'} />
            <MenuItem setActiveNavLink={setActiveNavLink} name={'experience'} />
            <MenuItem setActiveNavLink={setActiveNavLink} name={'education'} />
            <MenuItem setActiveNavLink={setActiveNavLink} name={'portfolio'} />
            <MenuItem setActiveNavLink={setActiveNavLink} name={'resume'} />
            <MenuItem setActiveNavLink={setActiveNavLink} name={'contact'} />
          </Nav>
          <Nav>
            <LocaleDropDown />
          </Nav>
        </Navbar.Collapse>
      </Navbar>
    </div>
  )
}
