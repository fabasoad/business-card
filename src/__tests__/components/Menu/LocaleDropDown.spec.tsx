/// <reference types="jest" />
import { shallow, ShallowWrapper } from 'enzyme'
import * as React from 'react'
import { LocaleDropDown } from '../../../components/Menu/LocaleDropDown'
import { Locale } from '../../../store/locale/types'

test('should render LocaleDropDown correctly', () => {
  const getLocalesExceptOfSpy = jest.fn(() => [{ code: 'gb', title: 'EN' }])
  const wrapper: ShallowWrapper = shallow(<LocaleDropDown
    getLocalesExceptOf={getLocalesExceptOfSpy}
    locale={{ code: 'ru', title: 'RU' }}
    startSetLocale={null}
  />)
  expect(wrapper).toMatchSnapshot()
  expect(getLocalesExceptOfSpy).toHaveBeenCalledTimes(1)
  expect(getLocalesExceptOfSpy).toHaveBeenCalledWith('ru')
})

test('should change locale correctly', () => {
  const expectedLocale: Locale = { code: 'ru', title: 'RU' }
  const startSetLocaleSpy = jest.fn()
  const wrapper: ShallowWrapper = shallow(<LocaleDropDown
    getLocalesExceptOf={() => [expectedLocale]}
    locale={{ code: 'ru', title: 'RU' }}
    startSetLocale={startSetLocaleSpy}
  />)
  wrapper.find('DropdownItem').simulate('click')
  expect(startSetLocaleSpy).toHaveBeenCalledTimes(1)
  expect(startSetLocaleSpy).toHaveBeenCalledWith(expectedLocale)
})
