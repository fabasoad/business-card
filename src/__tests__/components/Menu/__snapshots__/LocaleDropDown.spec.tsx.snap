// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LocaleDropDown correctly 1`] = `
<Dropdown
  navbar={false}
>
  <DropdownToggle
    bsPrefix="nav-link dropdown-toggle"
    id="btnLocale"
    variant={null}
  >
    <FlagIcon
      code="ru"
    />
    RU
  </DropdownToggle>
  <DropdownMenu
    align="left"
    alignRight={false}
    flip={true}
  >
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      bsPrefix="nav-link"
      disabled={false}
      eventKey="gb"
      key="gb"
      onSelect={[Function]}
    >
      <FlagIcon
        code="gb"
      />
      EN
    </DropdownItem>
  </DropdownMenu>
</Dropdown>
`;
